Recruit-Training/Advanced-Recruit-Training/src/LogFileReader.cpp

```diff
@@ -45,27 +47,29 @@ bool LogFileReader::readAll(const QString& fileName)
             // This line notifies the classes listening to this signal
             // that battery data has been received.
             emit batteryDataReceived(batteryData);
-        }
+         }
     }
 
     return true;
 }
 
-/* File input is a csv file in the format of hh:mm:ss:zzz, voltage, current.
- * Negative current values denote a charging battery.
- * Need to implement error checking for the correct number of values and
- * that the conversion from string to double is sucessful.*/
 bool LogFileReader::parseLine(const QString& line, BatteryData& batteryData) const
 {
-    QStringList sections = line.split(BATDATA_DELIMITER);
+    bool validVoltage_= 0;
+    bool validCurrent_= 0;
 
-    QString timeString = sections.at(0);
-    batteryData.time = QTime::fromString(timeString, STRING_TIME_FORMAT);
+    QStringList sections= line.split(BATDATA_DELIMITER);
 
-    batteryData.voltage = sections.at(1).toDouble();
+    QString timeString= sections.at(0);
+    batteryData.time= QTime::fromString(timeString, STRING_TIME_FORMAT);
 
-    batteryData.current = sections.at(2).toDouble();
+    batteryData.voltage= sections.at(1).toDouble(&validVoltage_);
+    batteryData.current= sections.at(2).toDouble(&validCurrent_);
 
-    return true;
+    if (sections.size()!= COLUMNS || validVoltage_== false || validCurrent_== false || batteryData.time.isValid()== false )
```

Comment:
You need to check the `sections.size()` first, otherwise you'll be checking it AFTER already using it at lines 66/67.

Instead of doing checks like `validVoltage_== false`, do `!validVoltage_`